src/hooks/useReplyBox.tsx

```diff
@@ -0,0 +1,284 @@
+import { createNote } from "@/services/wrike";
+import { query } from "@/utils/query";
+import { truncate } from "@/utils";
+import { APP_PREFIX } from "@/constants";
+import type { FC, PropsWithChildren } from "react";
+import type { IDeskproClient, GetStateResponse, TargetAction } from "@deskpro/app-sdk";
+import type { ITaskFromList } from "@/services/wrike/types";
+import type { TicketData } from "@/types";
+
+export type ReplyBoxType = "note" | "email";
+
+export type SetSelectionState = (taskId: ITaskFromList["id"], selected: boolean, type: ReplyBoxType) => void|Promise<{ isSuccess: boolean }|void>;
+
+export type GetSelectionState = (taskId: ITaskFromList["id"], type: ReplyBoxType) => void|Promise<Array<GetStateResponse<string>>>;
+
+export type DeleteSelectionState = (taskId: ITaskFromList["id"], type: ReplyBoxType) => void|Promise<boolean|void>;
+
+type ReturnUseReplyBox = {
+  setSelectionState: SetSelectionState,
+  getSelectionState: GetSelectionState,
+  deleteSelectionState: DeleteSelectionState,
+};
+
+const noteKey = (ticketId: string, taskId: ITaskFromList["id"]|"*") => {
+  return `tickets/${ticketId}/${APP_PREFIX}/notes/selection/${taskId}`.toLowerCase();
+};
+
+const emailKey = (ticketId: string, taskId: ITaskFromList["id"]|"*") => {
+  return `tickets/${ticketId}/${APP_PREFIX}/emails/selection/${taskId}`.toLowerCase();
+};
+
+const registerReplyBoxNotesAdditionsTargetAction = (
+  client: IDeskproClient,
+  ticketId: TicketData["ticket"]["id"],
+  taskIds: Array<ITaskFromList["id"]>,
+  tasksMap: Record<ITaskFromList["id"], ITaskFromList>,
+): void|Promise<void> => {
+  if (!ticketId) {
+    return;
+  }
+
+  if (Array.isArray(taskIds) && !taskIds.length) {
+    return client.deregisterTargetAction(`${APP_PREFIX}ReplyBoxNoteAdditions`);
+  }
+
+  return Promise
+    .all(taskIds.map((taskId: ITaskFromList["id"]) => client.getState<{ selected: boolean }>(noteKey(ticketId, taskId))))
+    .then((flags) => {
+      client.registerTargetAction(`${APP_PREFIX}ReplyBoxNoteAdditions`, "reply_box_note_item_selection", {
+        title: "Add to Wrike",
+        payload: taskIds.map((taskId, idx) => {
+          const task = tasksMap[taskId];
+          return {
+            id: task.id,
+            title: truncate(task.title),
+            selected: flags[idx][0]?.data?.selected ?? false,
+          };
+        }),
+      });
+    })
+    ;
+};
+
+const registerReplyBoxEmailsAdditionsTargetAction = (
+  client: IDeskproClient,
+  ticketId: TicketData["ticket"]["id"],
+  taskIds: Array<ITaskFromList["id"]>,
+  tasksMap: Record<ITaskFromList["id"], ITaskFromList>,
+): void|Promise<void> => {
+  if (!ticketId) {
+    return;
+  }
+
+  if (Array.isArray(taskIds) && !taskIds.length) {
+    return client.deregisterTargetAction(`${APP_PREFIX}ReplyBoxEmailAdditions`);
+  }
+
+  return Promise
+    .all(taskIds.map((taskId: ITaskFromList["id"]) => {
+      return client.getState<{ selected: boolean }>(emailKey(ticketId, taskId))
+    }))
```

Comment:
Could this be expensive if we get a lot of taskIds? Instead, can we wait for one to finish before getting the next?

```
const results = [];
for (const taskId of taskIds) {
    results.push(await client.getState<{ selected: boolean }>(emailKey(ticketId, taskId)))
}

// extra stuff

return results;
```

Reply:
Honestly, I doubt that many tasks can be linked to a ticket. We can optimize this later if needed. Or we can do it now if you think it’s important.

Reply:
Fair enough, happy to leave as is for now, but maybe leave a tech debt ticket or note somewhere saying that this should be revisited if clients often link many-many tickets.

Reply:
added comment in code